src/Symfony/Bundle/FrameworkBundle/Messenger/RouterContextMiddleware.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Messenger;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+use Symfony\Component\Messenger\Middleware\StackInterface;
+use Symfony\Component\Messenger\Stamp\ConsumedByWorkerStamp;
+use Symfony\Component\Routing\RequestContext;
+use Symfony\Component\Routing\RequestContextAwareInterface;
+
+/**
+ * Restore the Router context when processing the message.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RouterContextMiddleware implements MiddlewareInterface
+{
+    private $router;
+
+    public function __construct(RequestContextAwareInterface $router)
+    {
+        $this->router = $router;
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (null !== $envelope->last(ConsumedByWorkerStamp::class)) {
+            /** @var RouterContextStamp $contextStamp */
+            if (null !== $contextStamp = $envelope->last(RouterContextStamp::class)) {
+                $currentContext = $this->router->getContext();
+
+                $this->router->setContext(new RequestContext(
+                    $contextStamp->getBaseUrl(),
+                    $contextStamp->getMethod(),
+                    $contextStamp->getHost(),
+                    $contextStamp->getScheme(),
+                    $contextStamp->getHttpPort(),
+                    $contextStamp->getHttpsPort(),
+                    $contextStamp->getPathInfo(),
+                    $contextStamp->getQueryString()
+                ));
+
+                try {
+                    return $stack->next()->handle($envelope, $stack);
+                } finally {
+                    $this->router->setContext($currentContext);
+                }
+            }
+        } else {
+            $context = $this->router->getContext();
+            $envelope = $envelope->with(new RouterContextStamp(
+                $context->getBaseUrl(),
```

Comment:
why don't you pass the $context directly as a RequestContext object?

Reply:
because it contains closures (in `arguments`) which are not serializable.